src/util/integer_gmp_imp.cpp

```diff
@@ -340,6 +340,21 @@ bool Integer::isNegativeOne() const
 }
 
 Integer Integer::pow(unsigned long int exp) const
+{
+  // check if exp is within the uint32_t range
+  uint32_t low = static_cast<uint32_t>(exp);
+  uint32_t high = static_cast<uint32_t>(exp << 32);
+  if (low == exp) {
+    // if it is, safely call the gmp pow function
+    return powHelper(exp);
+  } else {
+      // if exp is bigger than max uint32_t, use the following identity:
+      // a^(x + 2^32*y) = a^x * (a^(2^32))^y
+      return powHelper(low) * (powHelper(static_cast<uint32_t>(-1))).powHelper(high);
```

Comment:
I **think** it is still not right.
Note that the `-1` corresponds to `2^32 - 1`, which is why we need to have `(powHelper(-1) * (*this)).powHelper(high)`.